Mathlib/SetTheory/Nimber/Field.lean

```diff
@@ -218,4 +222,124 @@ instance : CommRing Nimber where
+
+This preliminary definition "accidentally" satisfies `invAux 0 = 1`, which the real inverse
+corrects. The lemma `inv_eq_invAux` can be used to transfer between the two. -/
+def invAux (a : Nimber) : Nimber :=
+  sInf (invSet a)ᶜ
+termination_by (a, 1)
+
+/-- The set in the definition of `invAux a`. -/
+def invSet (a : Nimber) : Set Nimber :=
+  ⋂₀ {s | 0 ∈ s ∧ ∀ a' < a, a' ≠ 0 → ∀ b ∈ s, invAux a' * (1 + (a + a') * b) ∈ s}
+termination_by (a, 0)
+
+end
+
+theorem zero_mem_invSet (a : Nimber) : 0 ∈ invSet a := by
+  rw [invSet]
+  exact Set.mem_sInter.2 fun _ hs => hs.1
+
+/-- "cons" is our operation `(1 + (a + a') * b) / a'` in the definition of the inverse. -/
+theorem cons_mem_invSet {a' : Nimber} (ha₀ : a' ≠ 0) (ha : a' < a) (hb : b ∈ invSet a) :
+    invAux a' * (1 + (a + a') * b) ∈ invSet a := by
+  rw [invSet] at hb ⊢
+  exact Set.mem_sInter.2 fun _ hs => hs.2 _ ha ha₀ _ (Set.mem_sInter.1 hb _ hs)
+
+/-- A recursion principle for `invSet`. -/
+@[elab_as_elim]
+theorem invSet_recOn {p : Nimber → Prop} (a : Nimber) (h0 : p 0)
+    (hi : ∀ a' < a, a' ≠ 0 → ∀ b, p b → p (invAux a' * (1 + (a + a') * b))) (x : Nimber)
+    (hx : x ∈ invSet a) : p x := by
+  revert x
+  rw [invSet]
+  exact Set.sInter_subset_of_mem ⟨h0, hi⟩
+
+/-- An enumeration of elements in `invSet` by a type in the same universe. -/
+private def List.toNimber {a : Nimber} : List a.toOrdinal.toType → Nimber
+  | [] => 0
+  | x::l =>
+    let a' := Ordinal.toNimber ((Ordinal.enumIsoToType a.toOrdinal).symm x)
+    invAux a' * (1 + (a + a') * (toNimber l))
+
+instance (a : Nimber.{u}) : Small.{u} (invSet a) := by
+  refine @small_subset.{u, u + 1} _ _ _ ?_ (small_range (@List.toNimber a))
+  refine fun x hx ↦ invSet_recOn a ⟨[], rfl⟩ ?_ x hx
+  rintro a' ha _ _ ⟨l, rfl⟩
+  use (Ordinal.enumIsoToType _ ⟨toOrdinal a', ha⟩)::l
+  rw [List.toNimber]
+  simp
+
+/-- The complement of `invSet a` is nonempty. -/
+private theorem invSet_nonempty (a : Nimber) : (invSet a)ᶜ.Nonempty :=
+  have := instSmallElemInvSet a -- why is this needed?
+  nonempty_of_not_bddAbove (Ordinal.not_bddAbove_compl_of_small _)
+
+theorem invAux_ne_zero (a : Nimber) : invAux a ≠ 0 := by
+  rw [invAux]
+  exact fun h ↦ h ▸ csInf_mem (invSet_nonempty a) <| zero_mem_invSet a
+
+theorem mem_invSet_of_lt_invAux (h : b < invAux a) : b ∈ invSet a := by
+  rw [invAux] at h
+  have := not_mem_of_lt_csInf h ⟨_, bot_mem_lowerBounds _⟩
+  rwa [Set.not_mem_compl_iff] at this
+
+theorem invAux_not_mem_invSet (a : Nimber) : invAux a ∉ invSet a := by
+  rw [invAux]
+  exact csInf_mem (invSet_nonempty a)
+
+theorem invAux_mem_invSet_of_lt (ha : a ≠ 0) (hb : a < b) : invAux a ∈ invSet b := by
+  have H := cons_mem_invSet ha hb (zero_mem_invSet b)
+  rwa [mul_zero, add_zero, mul_one] at H
+
+/-- We set up a simultaneous induction to prove that `invAux a` is the inverse of `a`, and no
+element in its defining set `invSet a` is. -/
+private theorem mul_inv_cancel_aux (a : Nimber) :
```

Comment:
Why is this and `invSet_nonempty` private, but other theorems are not? Are `invAux` and `invSet` part of the public API or not? This looks good to go but right now it looks like `private` is being applied kind of randomly? What is the logic here?

Reply:
`invSet_nonempty` is private since it's implied by `invAux_mem_invSet`, and likewise this is subsumed by `inv_eq_invAux`.

I did intend for `invSet` and `invAux` to be public API. I'm not sure if this characterization is ever the easiest way to find a nimber's inverse, but I do think that it's a nice theorem worth exposing.